modules/ITK/cli/utils/statismo-build-gp-model-kernels.h

```diff
@@ -95,12 +95,181 @@ class GaussianKernel : public statismo::ScalarValuedKernel<TPoint> {
         return os.str();
     }
 
-  private:
+private:
     double m_sigma;
     double m_sigma2;
 };
 
 
+
+
+
+template <class TPoint>
+class BSplineKernel : public statismo::ScalarValuedKernel<TPoint> {
+public:
+    typedef typename  TPoint::CoordRepType CoordRepType;
+    typedef vnl_vector<CoordRepType> VectorType;
+
+
+    BSplineKernel(double support) : m_support(support) {}
+
+
+    // the 3rd order b-spline basis function
+    double bspline3(const double& x) const {
+        const double absX = std::fabs(x);
+        const double absXSquared = absX * absX;
+        const double absXCube = absXSquared * absX;
+        const double twoMinAbsX = 2.0 - absX;
+        const double twoMinAbsXCube = twoMinAbsX * twoMinAbsX * twoMinAbsX;
+        const double twoByThree = 2.0 / 3.0;
+
+        double splineValue = 0;
+        if (absX >= 0 && absX < 1) {
+            splineValue = twoByThree - absXSquared + 0.5 * absXCube;
+        }
+        else if (absX >= 1 && absX < 2) {
+            splineValue = twoMinAbsXCube / 6.0;
+        }
+        else {
+            splineValue = 0;
+        }
+        return splineValue;
+    }
+
+    double tensorProductSpline(const VectorType& x) const {
+        double prod = 1;
+        for (unsigned d = 0; d < x.size(); ++d) {
+            prod *= bspline3(x[d]);
+        }
+        return prod;
+    }
+
+
+    inline double operator()(const TPoint& x, const TPoint& y) const {
+
+        assert(x.Size() == y.Size());
+        const unsigned dim = x.Size();
```

Comment:
Maybe the cleanest solution would be to do a template specialization for each dimension of the B-Spline Kernel. This would also solve the loop problem above. 

As doing a refactoring of the kernel structure is anyway on the todo list, I propose to postpone this change.